### ISSUE TITLE
Inifinite loop during concretization

### ISSUE THREAD

**Issue by skosukhin** [spack/spack#6524]
There is a package `openjpeg` with the following versions:
```python
version('2.1',   '3e1c451c087f8462955426da38aa3b3d')
version('2.0.1', '105876ed43ff7dbb2f90b41b5a43cfa5')
```
Note that the version tuples have different lengths.

Now, there is a package `eccodes` that depends on `openjpeg`. Recently, I have learnt that it can be built only with `openjpeg` `1.5.x` or `2.1.x`. So, I decided to put the constraint on the version of `openjpeg`: `1.5:1.5.999,2.1:2.1.999`. And got an infinie loop while doing `spack spec` or `spac install` here: https://github.com/spack/spack/blob/3dafbe901a2a2ade232f3221dd27786ee6f42097/lib/spack/spack/spec.py#L1793-L1798

The problem remains even if we specify just `depends_on('openjpeg@2.1:2.1.999')`. What solves the problem is making all the version tuples of `openjpeg` to have the same length, which I plan to do for now.

**Comment by healther** [comment]
Is this issue resolved?

**Comment by alalazo** [comment]
Closing. This currently errors out with the original concretizer and is resolved correctly by #19501
